File: web/src/data/bengaliCalendar.ts
export interface BengaliSeason {
  nameEn: string;
  nameBn: string;
  descriptionEn: string;
  descriptionBn: string;
}

export interface BengaliDateInfo {
  day: number;
  dayBn: string;
  monthNameEn: string;
  monthNameBn: string;
  year: number;
  yearBn: string;
  season: BengaliSeason;
  formattedBn: string;
  formattedEn: string;
}

export function toBengaliDigits(input: string | number): string {
  const bengaliDigits = ["০", "১", "২", "৩", "৪", "৫", "৬", "৭", "৮", "৯"];
  return String(input).replace(/[0-9]/g, (w) => bengaliDigits[parseInt(w, 10)]);
}

export function getBengaliDate(dateInput: Date = new Date()): BengaliDateInfo {
  const d = new Date(dateInput);
  const year = d.getFullYear();
  const month = d.getMonth() + 1; // 1-12
  const day = d.getDate();

  const isLeap = (year % 4 === 0 && year % 100 !== 0) || year % 400 === 0;

  // Bengali Academy revised rules
  let bYear = year - 593;
  if (month < 4 || (month === 4 && day < 14)) {
    bYear -= 1;
  }

  let bMonthIndex = 0; // 0: Boishakh .. 11: Choitra
  let bDay = 1;

  if (month === 4 && day >= 14) {
    bMonthIndex = 0;
    bDay = day - 14 + 1;
  } else if (month === 5) {
    if (day <= 14) {
      bMonthIndex = 0;
      bDay = 17 + day;
    } else {
      bMonthIndex = 1;
      bDay = day - 14;
    }
  } else if (month === 6) {
    if (day <= 14) {
      bMonthIndex = 1;
      bDay = 17 + day;
    } else {
      bMonthIndex = 2;
      bDay = day - 14;
    }
  } else if (month === 7) {
    if (day <= 15) {
      bMonthIndex = 2;
      bDay = 16 + day;
    } else {
      bMonthIndex = 3;
      bDay = day - 15;
    }
  } else if (month === 8) {
    if (day <= 15) {
      bMonthIndex = 3;
      bDay = 16 + day;
    } else {
      bMonthIndex = 4;
      bDay = day - 15;
    }
  } else if (month === 9) {
    if (day <= 15) {
      bMonthIndex = 4;
      bDay = 16 + day;
    } else {
      bMonthIndex = 5;
      bDay = day - 15;
    }
  } else if (month === 10) {
    if (day <= 15) {
      bMonthIndex = 5;
      bDay = 15 + day;
    } else {
      bMonthIndex = 6;
      bDay = day - 15;
    }
  } else if (month === 11) {
    if (day <= 14) {
      bMonthIndex = 6;
      bDay = 16 + day;
    } else {
      bMonthIndex = 7;
      bDay = day - 14;
    }
  } else if (month === 12) {
    if (day <= 14) {
      bMonthIndex = 7;
      bDay = 16 + day;
    } else {
      bMonthIndex = 8;
      bDay = day - 14;
    }
  } else if (month === 1) {
    if (day <= 13) {
      bMonthIndex = 8;
      bDay = 17 + day;
    } else {
      bMonthIndex = 9;
      bDay = day - 13;
    }
  } else if (month === 2) {
    if (day <= 12) {
      bMonthIndex = 9;
      bDay = 18 + day;
    } else {
      bMonthIndex = 10;
      bDay = day - 12;
    }
  } else if (month === 3) {
    const falgunDays = isLeap ? 30 : 29;
    const cutOff = falgunDays - 16;
    if (day <= cutOff) {
      bMonthIndex = 10;
      bDay = 16 + day;
    } else {
      bMonthIndex = 11;
      bDay = day - cutOff;
    }
  } else if (month === 4 && day < 14) {
    bMonthIndex = 11;
    bDay = day + 17;
  }

  const MONTH_NAMES_EN = [
    "Boishakh", "Joishtho", "Asharh", "Shrabon", "Bhadro", "Ashwin",
    "Kartik", "Agrahayan", "Poush", "Magh", "Falgun", "Choitra"
  ];
  const MONTH_NAMES_BN = [
    "বৈশাখ", "জ্যৈষ্ঠ", "আষাঢ়", "শ্রাবণ", "ভাদ্র", "আশ্বিন",
    "কার্তিক", "অগ্রহায়ণ", "পৌষ", "মাঘ", "ফাল্গুন", "চৈত্র"
  ];

  const SEASONS: BengaliSeason[] = [
    { nameEn: "Grishma (Summer)", nameBn: "গ্রীষ্ম কাল", descriptionEn: "Intense heat and Kalbaishakhi storms", descriptionBn: "তীব্র তাপদাহ ও কালবৈশাখীর কাল" },
    { nameEn: "Barsha (Monsoon)", nameBn: "বর্ষা কাল", descriptionEn: "Heavy monsoon rains and lush wetlands", descriptionBn: "টানা বর্ষণ ও নদ-নদীতে প্লাবনের কাল" },
    { nameEn: "Sharat (Autumn)", nameBn: "শরৎ কাল", descriptionEn: "Clear blue skies and blooming kashful", descriptionBn: "শুভ্র মেঘ ও কাশফুলের শোভা" },
    { nameEn: "Hemanta (Late Autumn)", nameBn: "হেমন্ত কাল", descriptionEn: "Golden paddy harvest and early mist", descriptionBn: "নবান্ন ও সোনালী ধানের উৎসব" },
    { nameEn: "Sheet (Winter)", nameBn: "শীত কাল", descriptionEn: "Morning river fog and date palm sap", descriptionBn: "কুয়াশাচ্ছন্ন সকাল ও খেজুর রসের মিষ্টি আমেজ" },
    { nameEn: "Basanta (Spring)", nameBn: "বসন্ত কাল", descriptionEn: "Blooming nature and gentle southern breeze", descriptionBn: "দক্ষিণা বাতাস ও ফুল ফোটার ঋতু" }
  ];

  const seasonIndex = Math.floor(bMonthIndex / 2);
  const season = SEASONS[seasonIndex] || SEASONS[0];

  const dayBn = toBengaliDigits(bDay);
  const yearBn = toBengaliDigits(bYear);

  return {
    day: bDay,
    dayBn,
    monthNameEn: MONTH_NAMES_EN[bMonthIndex],
    monthNameBn: MONTH_NAMES_BN[bMonthIndex],
    year: bYear,
    yearBn,
    season,
    formattedBn: `${dayBn} ${MONTH_NAMES_BN[bMonthIndex]}, ${yearBn} বঙ্গাব্দ`,
    formattedEn: `${bDay} ${MONTH_NAMES_EN[bMonthIndex]}, ${bYear} Bangabda`
  };
}

export function calculatePrayerTimes(lat: number, lon: number, date: Date = new Date()) {
  const dayOfYear = Math.floor((date.getTime() - new Date(date.getFullYear(), 0, 0).getTime()) / 1000 / 60 / 60 / 24);
  const d = dayOfYear;
  const b = (2 * Math.PI * (d - 81)) / 365.0;
  const eot = 9.87 * Math.sin(2 * b) - 7.53 * Math.cos(b) - 1.5 * Math.sin(b);
  const delta = (23.45 * Math.PI / 180) * Math.sin((2 * Math.PI * (d - 81)) / 365.0);
  const latRad = (lat * Math.PI) / 180;
  const solarNoon = 12.0 - (lon - 6.0 * 15.0) / 15.0 - eot / 60.0;

  function hourAngle(altitudeDeg: number): number {
    const altRad = (altitudeDeg * Math.PI) / 180;
    const cosH = (Math.sin(altRad) - Math.sin(latRad) * Math.sin(delta)) / (Math.cos(latRad) * Math.cos(delta));
    const clamped = Math.max(-1.0, Math.min(1.0, cosH));
    return (Math.acos(clamped) * 180 / Math.PI) / 15.0;
  }

  const hSunrise = hourAngle(-0.833);
  const hFajr = hourAngle(-18.0);
  const hIsha = hourAngle(-18.0);

  const shadowFactor = 1.5;
  const noonAltRad = Math.PI / 2.0 - Math.abs(latRad - delta);
  const asrAltRad = Math.atan(1.0 / (shadowFactor + 1.0 / Math.tan(noonAltRad)));
  const hAsr = hourAngle((asrAltRad * 180) / Math.PI);

  const fajrTime = solarNoon - hFajr;
  const sunriseTime = solarNoon - hSunrise;
  const dhuhrTime = solarNoon + 4.0 / 60.0;
  const asrTime = solarNoon + hAsr;
  const maghribTime = solarNoon + hSunrise + 3.0 / 60.0;
  const ishaTime = solarNoon + hIsha;

  function fmt(hoursDec: number): string {
    const mins = Math.round(hoursDec * 60);
    let h = Math.floor(mins / 60) % 24;
    const m = mins % 60;
    if (h < 0) h += 24;
    return `${h.toString().padStart(2, "0")}:${m.toString().padStart(2, "0")}`;
  }

  return {
    fajr: fmt(fajrTime),
    sunrise: fmt(sunriseTime),
    dhuhr: fmt(dhuhrTime),
    asr: fmt(asrTime),
    maghrib: fmt(maghribTime),
    isha: fmt(ishaTime),
  };
}

File: app/src/main/java/com/example/data/model/BengaliCalendar.kt
package com.example.data.model

import java.util.Calendar
import java.util.TimeZone

enum class BengaliSeason(
    val nameEn: String,
    val nameBn: String,
    val descriptionEn: String,
    val descriptionBn: String,
    val iconKey: String
) {
    GRISHMO("Grishmo (Summer)", "গ্রীষ্ম কাল", "High heat, humid winds & Kalbaishakhi thunderstorms", "তপ্ত রোদ, কালবৈশাখী ও পাকা ফলের সুবাস", "summer"),
    BARSHA("Barsha (Monsoon)", "বর্ষা কাল", "Heavy monsoon rains, active river flows & cloudbursts", "মেঘভাঙা বারিধারা, সজল বাতাস ও প্লাবিত নদী", "monsoon"),
    SHARAT("Sharat (Autumn)", "শরৎ কাল", "Clear blue skies, drifting white kashful clouds", "নীল আকাশে সাদা মেঘের ভেলা ও শুভ্র কাশফুল", "autumn"),
    HEMANTO("Hemanto (Late Autumn)", "হেমন্ত কাল", "Dewy dawns, golden paddy harvest & gentle chill", "ভোরের শিশিরভেজা ঘাস, পাকা আমন ধান ও নবান্ন", "late_autumn"),
    SHEET("Sheet (Winter)", "শীত কাল", "Cool northern breeze, morning mist & date palm sap", "উত্তুরে হাওয়া, ভোরের কুয়াশা ও খেজুরের রস", "winter"),
    BOSHONTO("Boshonto (Spring)", "বসন্ত কাল", "Pleasant warmth, blooming shimul & south breeze", "দক্ষিণা সমীরণ, পলাশ-শিমুল ও কোকিলের গান", "spring")
}

data class BengaliDateInfo(
    val dayBn: String,
    val dayEn: Int,
    val monthBn: String,
    val monthEn: String,
    val yearBn: String,
    val yearEn: Int,
    val formattedBn: String,
    val formattedEn: String,
    val season: BengaliSeason
)

object BengaliCalendarHelper {

    private val BANGLA_DIGITS = charArrayOf('০', '১', '২', '৩', '৪', '৫', '৬', '৭', '৮', '৯')

    fun toBengaliDigits(input: Any): String {
        val str = input.toString()
        val sb = java.lang.StringBuilder()
        for (ch in str) {
            if (ch in '0'..'9') {
                sb.append(BANGLA_DIGITS[ch - '0'])
            } else {
                sb.append(ch)
            }
        }
        return sb.toString()
    }

    /**
     * Converts a Gregorian timestamp (or now) to the official Bangladesh Bengali Academy calendar
     */
    fun getBengaliDate(timestamp: Long = System.currentTimeMillis()): BengaliDateInfo {
        val cal = Calendar.getInstance(TimeZone.getTimeZone("Asia/Dhaka"))
        cal.timeInMillis = timestamp

        val gYear = cal.get(Calendar.YEAR)
        val gMonth = cal.get(Calendar.MONTH) + 1 // 1-12
        val gDay = cal.get(Calendar.DAY_OF_MONTH)

        // Bangladesh standard Bengali calendar starts on April 14
        val isLeapYear = (gYear % 4 == 0 && gYear % 100 != 0) || (gYear % 400 == 0)

        // Bengali month names
        val monthsBn = listOf("বৈশাখ", "জ্যৈষ্ঠ", "আষাঢ়", "শ্রাবণ", "ভাদ্র", "আশ্বিন", "কার্তিক", "অগ্রহায়ণ", "পৌষ", "মাঘ", "ফাল্গুন", "চৈত্র")
        val monthsEn = listOf("Boishakh", "Joishtho", "Asharh", "Srabon", "Bhadro", "Ashwin", "Kartik", "Ograhayon", "Poush", "Magh", "Falgun", "Choitro")

        var bMonthIdx: Int
        var bDay: Int
        var bYear: Int = if (gMonth > 4 || (gMonth == 4 && gDay >= 14)) gYear - 593 else gYear - 594

        when {
            // April 14 - May 14 : Boishakh (31 days)
            gMonth == 4 && gDay >= 14 -> {
                bMonthIdx = 0
                bDay = gDay - 13
            }
            gMonth == 5 && gDay <= 14 -> {
                bMonthIdx = 0
                bDay = gDay + 17
            }
            // May 15 - June 14 : Joishtho (31 days)
            gMonth == 5 && gDay >= 15 -> {
                bMonthIdx = 1
                bDay = gDay - 14
            }
            gMonth == 6 && gDay <= 14 -> {
                bMonthIdx = 1
                bDay = gDay + 17
            }
            // June 15 - July 15 : Asharh (31 days)
            gMonth == 6 && gDay >= 15 -> {
                bMonthIdx = 2
                bDay = gDay - 14
            }
            gMonth == 7 && gDay <= 15 -> {
                bMonthIdx = 2
                bDay = gDay + 16
            }
            // July 16 - August 15 : Srabon (31 days)
            gMonth == 7 && gDay >= 16 -> {
                bMonthIdx = 3
                bDay = gDay - 15
            }
            gMonth == 8 && gDay <= 15 -> {
                bMonthIdx = 3
                bDay = gDay + 16
            }
            // August 16 - September 15 : Bhadro (31 days)
            gMonth == 8 && gDay >= 16 -> {
                bMonthIdx = 4
                bDay = gDay - 15
            }
            gMonth == 9 && gDay <= 15 -> {
                bMonthIdx = 4
                bDay = gDay + 16
            }
            // September 16 - October 15 : Ashwin (31 days)
            gMonth == 9 && gDay >= 16 -> {
                bMonthIdx = 5
                bDay = gDay - 15
            }
            gMonth == 10 && gDay <= 15 -> {
                bMonthIdx = 5
                bDay = gDay + 15
            }
            // October 16 - November 14 : Kartik (30 days)
            gMonth == 10 && gDay >= 16 -> {
                bMonthIdx = 6
                bDay = gDay - 15
            }
            gMonth == 11 && gDay <= 14 -> {
                bMonthIdx = 6
                bDay = gDay + 16
            }
            // November 15 - December 14 : Ograhayon (30 days)
            gMonth == 11 && gDay >= 15 -> {
                bMonthIdx = 7
                bDay = gDay - 14
            }
            gMonth == 12 && gDay <= 14 -> {
                bMonthIdx = 7
                bDay = gDay + 16
            }
            // December 15 - January 13 : Poush (30 days)
            gMonth == 12 && gDay >= 15 -> {
                bMonthIdx = 8
                bDay = gDay - 14
            }
            gMonth == 1 && gDay <= 13 -> {
                bMonthIdx = 8
                bDay = gDay + 17
            }
            // January 14 - February 12 : Magh (30 days)
            gMonth == 1 && gDay >= 14 -> {
                bMonthIdx = 9
                bDay = gDay - 13
            }
            gMonth == 2 && gDay <= 12 -> {
                bMonthIdx = 9
                bDay = gDay + 18
            }
            // February 13 - March 14 : Falgun (29 or 30 days)
            gMonth == 2 && gDay >= 13 -> {
                bMonthIdx = 10
                bDay = gDay - 12
            }
            gMonth == 3 && gDay <= 14 -> {
                bMonthIdx = 10
                val febDays = if (isLeapYear) 29 else 28
                bDay = gDay + (febDays - 12)
            }
            // March 15 - April 13 : Choitro (30 days)
            else -> {
                bMonthIdx = 11
                bDay = if (gMonth == 3) gDay - 14 else gDay + 17
            }
        }

        val monthBn = monthsBn[bMonthIdx]
        val monthEn = monthsEn[bMonthIdx]

        // 6 Bengali Seasons: 2 months each
        val season = when (bMonthIdx) {
            0, 1 -> BengaliSeason.GRISHMO   // Boishakh, Joishtho
            2, 3 -> BengaliSeason.BARSHA    // Asharh, Srabon
            4, 5 -> BengaliSeason.SHARAT    // Bhadro, Ashwin
            6, 7 -> BengaliSeason.HEMANTO   // Kartik, Ograhayon
            8, 9 -> BengaliSeason.SHEET     // Poush, Magh
            else -> BengaliSeason.BOSHONTO  // Falgun, Choitro
        }

        val dayBn = toBengaliDigits(bDay)
        val yearBn = toBengaliDigits(bYear)

        val formattedBn = "$dayBn $monthBn $yearBn বঙ্গাব্দ"
        val formattedEn = "$bDay $monthEn $bYear Banglabda"

        return BengaliDateInfo(
            dayBn = dayBn,
            dayEn = bDay,
            monthBn = monthBn,
            monthEn = monthEn,
            yearBn = yearBn,
            yearEn = bYear,
            formattedBn = formattedBn,
            formattedEn = formattedEn,
            season = season
        )
    }
}
